src/shims/sync.rs

```diff
@@ -168,105 +138,160 @@ fn mutex_get_or_create_blockset<'mir, 'tcx: 'mir>(
+    set_at_offset(ecx, cond_op, 4, id, ecx.machine.layouts.u32, PTHREAD_COND_T_MIN_SIZE)
 }
 
-fn rwlock_get_or_create_writer_blockset<'mir, 'tcx: 'mir>(
+fn cond_get_or_create_id<'mir, 'tcx: 'mir>(
     ecx: &mut MiriEvalContext<'mir, 'tcx>,
-    rwlock_op: OpTy<'tcx, Tag>,
-) -> InterpResult<'tcx, BlockSetId> {
-    let blockset = rwlock_get_writer_blockset(ecx, rwlock_op)?.to_u32()?;
-    if blockset == 0 {
-        // 0 is a default value and also not a valid blockset id. Need to
-        // allocate a new blockset.
-        let blockset = ecx.create_blockset()?;
-        rwlock_set_writer_blockset(ecx, rwlock_op, blockset.to_u32_scalar())?;
-        Ok(blockset)
+    cond_op: OpTy<'tcx, Tag>,
+) -> InterpResult<'tcx, CondvarId> {
+    let id = cond_get_id(ecx, cond_op)?.to_u32()?;
+    if id == 0 {
+        // 0 is a default value and also not a valid conditional variable id.
+        // Need to allocate a new id.
+        let id = ecx.condvar_create();
+        cond_set_id(ecx, cond_op, id.to_u32_scalar())?;
+        Ok(id)
     } else {
-        Ok(BlockSetId::new(blockset))
+        Ok(CondvarId::from_u32(id))
     }
 }
 
-fn rwlock_get_reader_blockset<'mir, 'tcx: 'mir>(
+fn cond_get_clock_id<'mir, 'tcx: 'mir>(
     ecx: &MiriEvalContext<'mir, 'tcx>,
-    rwlock_op: OpTy<'tcx, Tag>,
+    cond_op: OpTy<'tcx, Tag>,
 ) -> InterpResult<'tcx, ScalarMaybeUndef<Tag>> {
-    get_at_offset(ecx, rwlock_op, 16, ecx.machine.layouts.u32, PTHREAD_RWLOCK_T_MIN_SIZE)
+    get_at_offset(ecx, cond_op, 8, ecx.machine.layouts.i32, PTHREAD_COND_T_MIN_SIZE)
 }
 
-fn rwlock_set_reader_blockset<'mir, 'tcx: 'mir>(
+fn cond_set_clock_id<'mir, 'tcx: 'mir>(
     ecx: &mut MiriEvalContext<'mir, 'tcx>,
-    rwlock_op: OpTy<'tcx, Tag>,
-    blockset: impl Into<ScalarMaybeUndef<Tag>>,
+    cond_op: OpTy<'tcx, Tag>,
+    clock_id: impl Into<ScalarMaybeUndef<Tag>>,
 ) -> InterpResult<'tcx, ()> {
-    set_at_offset(ecx, rwlock_op, 16, blockset, ecx.machine.layouts.u32, PTHREAD_RWLOCK_T_MIN_SIZE)
+    set_at_offset(ecx, cond_op, 8, clock_id, ecx.machine.layouts.i32, PTHREAD_COND_T_MIN_SIZE)
 }
 
-fn rwlock_get_or_create_reader_blockset<'mir, 'tcx: 'mir>(
+/// Try to reacquire the mutex associated with the condition variable after we were signaled.
+fn reacquire_cond_mutex<'mir, 'tcx: 'mir>(
     ecx: &mut MiriEvalContext<'mir, 'tcx>,
-    rwlock_op: OpTy<'tcx, Tag>,
-) -> InterpResult<'tcx, BlockSetId> {
-    let blockset = rwlock_get_reader_blockset(ecx, rwlock_op)?.to_u32()?;
-    if blockset == 0 {
-        // 0 is a default value and also not a valid blockset id. Need to
-        // allocate a new blockset.
-        let blockset = ecx.create_blockset()?;
-        rwlock_set_reader_blockset(ecx, rwlock_op, blockset.to_u32_scalar())?;
-        Ok(blockset)
+    thread: ThreadId,
+    mutex: MutexId,
+) -> InterpResult<'tcx> {
+    if ecx.mutex_is_locked(mutex) {
+        ecx.mutex_enqueue(mutex, thread);
```

Comment:
Seems reasonable to have an assertion here or even in `enqueue` that the thread must be currently blocked.